state/modelcredential_test.go

```diff
@@ -87,6 +88,137 @@ func (s *ModelCredentialSuite) TestValidateCloudCredentialModel(c *gc.C) {
+		cloud.NewCredential(cloud.EmptyAuthType, nil),
+	)
+}
+
+func (s *ModelCredentialSuite) TestSetCloudCredentialNoUpdate(c *gc.C) {
+	tag := names.NewCloudCredentialTag("dummy/bob/foobar")
+	m := s.assertSetCloudCredential(c,
+		tag,
+		cloud.NewCredential(cloud.EmptyAuthType, nil),
+	)
+
+	set, err := m.SetCloudCredential(tag)
+	c.Assert(err, jc.ErrorIsNil)
+	// This should be false as cloud credential change was an no-op.
+	c.Assert(set, jc.IsFalse)
+
+	// Check credential is still set.
+	credentialTag, credentialSet := m.CloudCredential()
+	c.Assert(credentialTag, gc.DeepEquals, tag)
+	c.Assert(credentialSet, jc.IsTrue)
+}
+
+func (s *ModelCredentialSuite) TestSetCloudCredentialInvalidCredentialContent(c *gc.C) {
+	tag := names.NewCloudCredentialTag("dummy/bob/foobar")
+	credential := cloud.NewCredential(cloud.EmptyAuthType, nil)
+	err := s.State.UpdateCloudCredential(tag, credential)
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.State.InvalidateCloudCredential(tag, "test")
+	c.Assert(err, jc.ErrorIsNil)
+
+	m, err := s.State.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	set, err := m.SetCloudCredential(tag)
+	c.Assert(err, gc.ErrorMatches, `credential "dummy/bob/foobar" not valid`)
+	c.Assert(set, jc.IsFalse)
+
+	credentialTag, credentialSet := m.CloudCredential()
+	// Make sure no credential is set.
+	c.Assert(credentialTag, gc.DeepEquals, names.CloudCredentialTag{})
+	c.Assert(credentialSet, jc.IsFalse)
+}
+
+func (s *ModelCredentialSuite) TestSetCloudCredentialInvalidCredentialForModel(c *gc.C) {
+	err := s.State.AddCloud(lowCloud, s.Owner.Name())
+	c.Assert(err, jc.ErrorIsNil)
+	tag := names.NewCloudCredentialTag("stratus/bob/foobar")
+	credential := cloud.NewCredential(cloud.AccessKeyAuthType, map[string]string{
+		"access-key": "someverysecretaccesskey",
+		"secret-key": "someverysercretplainkey",
+	})
+	err = s.State.UpdateCloudCredential(tag, credential)
+	c.Assert(err, jc.ErrorIsNil)
+
+	m, err := s.State.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	set, err := m.SetCloudCredential(tag)
+	c.Assert(err, gc.ErrorMatches, `cloud "stratus" not valid`)
+	c.Assert(set, jc.IsFalse)
+
+	credentialTag, credentialSet := m.CloudCredential()
+	// Make sure no credential is set.
+	c.Assert(credentialTag, gc.DeepEquals, names.CloudCredentialTag{})
+	c.Assert(credentialSet, jc.IsFalse)
+}
+
+func (s *ModelCredentialSuite) TestWatchModelCredential(c *gc.C) {
+	// Credential to use in this test.
+	tag := names.NewCloudCredentialTag("dummy/bob/foobar")
+	credential := cloud.NewCredential(cloud.EmptyAuthType, nil)
+	err := s.State.UpdateCloudCredential(tag, credential)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Model with credential watcher for this test.
+	m, err := s.State.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	w := m.WatchModelCredential()
+	defer statetesting.AssertStop(c, w)
+	wc := statetesting.NewNotifyWatcherC(c, s.State, w)
+
+	// Initial event.
+	wc.AssertOneChange()
+	wc.AssertNoChange()
+
+	// Check the watcher reacts to credential reference changes.
```

Comment:
Can we also have a check that setting a different attribute on the watched model does not trigger the credential watcher

Reply:
Yes, this is done in the same test - we are changing model's life and seeing no changes.